### DISCUSSION TITLE
Enable setting a label to an existing prompt version via the api/sdk.

### DISCUSSION THREAD

**Discussion by ClementNguyen** [langfuse/langfuse#2810] [Ideas]
### Describe the feature or potential improvement

Currently, labels can only be set to a prompt version via the API/SDK when creating a new version. The only way to set a label for an existing version is through the dashboard. This feature request proposes enabling the ability to set labels for existing prompt versions through the API/SDK.

### Additional information

This feature would be particularly useful for automating workflow processes. For example, in a CI/CD pipeline, there's currently no way to programmatically update all prompt versions with the 'dev' label to 'production'.

**Comment by cosasha97** [comment]
Would be super useful for me too!

**Comment by marcklingen** [comment]
Agree, thanks for sharing!

**Comment by marcklingen** [comment]
Additional notes on this by @mrdrprofuroboros (#3436)

> I have a bunch of prompts for my agent and I use version labels to mark the environment they are used in. And once I've done modifying prompts in 'development' I want to promote them all to 'staging' or 'production' in bulk. So far I can either go to the UI and click labels manually or create new versions with appropriate tag.
>
> That would be perfect if you could add update_labels or some similar API in PromptsClient.
>
> The pattern of usage is "set label X on latest version of the prompt NAME". Or more general "set label X on the prompt NAME's version where the label Y is" - that way we'd be able to promote 'production' to 'staging' even if 'latest'/'development' is way ahead

**Comment by shreyanair-ad** [comment]
Any update on this request?

**Reply by marcklingen** [reply]
Not yet, high up on the list of priorities of @hassiebp, open for contributions on this in the meantime

**Comment by eric8607242** [comment]
Hi all,

Here is my temporary solution for updating the labels. I hope this helps you as well.

```python
import psycopg2

def create_db_conn():
    conn_params = {
        'dbname': 'postgres',
        'user': 'postgres',
        'password': 'postgres',
        'host': '0.0.0.0',
        'port': 5432
    }
    return psycopg2.connect(**conn_params)

def assign_new_labels(conn, name, version, new_labels):
    try:
        with conn.cursor() as cursor:
            find_id_query = """
            SELECT id FROM prompts
            WHERE name = %s AND version = %s
            LIMIT 1
            """
            cursor.execute(find_id_query, (name, version))
            result = cursor.fetchone()
            
            if result is None:
                print("No row found with the specified name and version.")
                return
            
            target_id = result[0]

            remove_query = """
            UPDATE prompts
            SET labels = array(SELECT unnest(labels) EXCEPT SELECT unnest(%s::text[]))
            WHERE name = (SELECT name FROM prompts WHERE id = %s)
            AND labels && %s::text[]
            """
            cursor.execute(remove_query, (new_labels, target_id, new_labels))

            add_query = """
            UPDATE prompts
            SET labels = array(SELECT DISTINCT unnest(array_cat(labels, %s::text[])))
            WHERE id = %s
            """
            cursor.execute(add_query, (new_labels, target_id))

        conn.commit()
        print("Labels assigned successfully.")

    except Exception as e:
        conn.rollback()
        print(f"An error occurred: {e}")

conn = create_db_conn()
try:
    assign_production_label(conn, name='test', version=3, new_labels=["production", "test"])
finally:
    conn.close()
```

**Comment by hassiebp** [comment]
This has been released. You can now update prompt labels via the [API](https://api.reference.langfuse.com/#tag/promptversion/PATCH/api/public/v2/prompts/{name}/versions/{version}) and our SDK via the [api wrapper on the Langfuse client object](https://langfuse.com/changelog/2025-02-05-public-api-wrapper-sdks).

**Reply by igonro** [reply]
@marcklingen @hassiebp Is it just me or it is impossible to remove a label? It looks like I'm only able to add labels 😅 I have a label called "candidate" that I use for side-by-side responses, but after "competition" finishes I want to remove the "candidate" label.

**Reply by hassiebp** [reply]
Hi @igonro - indeed one can currently only append to the list of labels: https://api.reference.langfuse.com/#tag/promptversion/patch/api/public/v2/prompts/{name}/versions/{version} 

We will track the progress on allowing to remove labels via API here: https://github.com/langfuse/langfuse/issues/9342

**Reply by igonro** [reply]
Thanks @hassiebp I think now issues are duplicated #9278

Feel free to close my issue if needed.

**Reply by hassiebp** [reply]
@igonro - thanks, I closed my issue instead :)
